File: app/src/main/java/com/reteno/sample/testscreens/ScreenAdapter.java
package com.reteno.sample.testscreens;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.navigation.NavDirections;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.reteno.sample.databinding.ItemButtonGoToFragmentBinding;

import java.util.List;

public class ScreenAdapter extends ListAdapter<ScreenItem, ScreenAdapter.ScreenViewHolder> {

    private final ScreenItemClick itemClick;

    public ScreenAdapter(List<ScreenItem> screens, ScreenItemClick itemClick) {
        super(DIFF_CALLBACK);
        this.itemClick = itemClick;
        submitList(screens);
    }

    @NonNull
    @Override
    public ScreenViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ScreenViewHolder(
                ItemButtonGoToFragmentBinding.inflate(
                        LayoutInflater.from(parent.getContext()),
                        parent,
                        false
                ),
                itemClick);
    }

    @Override
    public void onBindViewHolder(ScreenViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    static class ScreenViewHolder extends RecyclerView.ViewHolder {

        private final ItemButtonGoToFragmentBinding binding;
        private final ScreenItemClick itemClick;

        ScreenViewHolder(ItemButtonGoToFragmentBinding binding, ScreenItemClick itemClick) {
            super(binding.getRoot());
            this.binding = binding;
            this.itemClick = itemClick;
        }

        public void bind(ScreenItem screen) {
            binding.buttonItem.setText(screen.getName());

            binding.buttonItem.setOnClickListener(v -> {
                if (screen.getNavigationId() != -1) {
                    if (screen.getBundle() != null) {
                        itemClick.navigateById(screen.getNavigationId(), screen.getBundle());
                    } else {
                        itemClick.navigateById(screen.getNavigationId());
                    }
                } else if (screen.getDirection() != null) {
                    itemClick.navigateByDirections(screen.getDirection());
                }
            });

        }
    }

    public static final DiffUtil.ItemCallback<ScreenItem> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<ScreenItem>() {
                @Override
                public boolean areItemsTheSame(
                        @NonNull ScreenItem oldItem, @NonNull ScreenItem newItem) {
                    return oldItem.getName().equals(newItem.getName());
                }

                @Override
                public boolean areContentsTheSame(
                        @NonNull ScreenItem oldItem, @NonNull ScreenItem newItem) {
                    return oldItem.equals(newItem);
                }
            };

    public interface ScreenItemClick {

        void navigateById(int fragmentId);

        void navigateById(int fragmentId, Bundle bundle);

        void navigateByDirections(NavDirections navDirections);
    }
}

File: app/src/main/java/com/reteno/sample/testscreens/ScreenAdapter.kt
package com.reteno.sample.testscreens

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.navigation.NavDirections
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.reteno.sample.databinding.ItemButtonGoToFragmentBinding
import com.reteno.sample.testscreens.ScreenAdapter.ScreenViewHolder

class ScreenAdapter(screens: List<ScreenItem>?, private val itemClick: ScreenItemClick) :
    ListAdapter<ScreenItem, ScreenViewHolder>(DIFF_CALLBACK) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ScreenViewHolder {
        return ScreenViewHolder(
            ItemButtonGoToFragmentBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            ),
            itemClick
        )
    }

    override fun onBindViewHolder(holder: ScreenViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    class ScreenViewHolder(
        private val binding: ItemButtonGoToFragmentBinding,
        private val itemClick: ScreenItemClick
    ) : RecyclerView.ViewHolder(
        binding.getRoot()
    ) {
        fun bind(screen: ScreenItem) {
            binding.buttonItem.text = screen.name
            binding.buttonItem.setOnClickListener {
                if (screen.navigationId != -1) {
                    if (screen.bundle != null) {
                        itemClick.navigateById(screen.navigationId, screen.bundle)
                    } else {
                        itemClick.navigateById(screen.navigationId)
                    }
                } else if (screen.direction != null) {
                    itemClick.navigateByDirections(screen.direction)
                }
            }
        }
    }

    init {
        submitList(screens)
    }

    interface ScreenItemClick {
        fun navigateById(fragmentId: Int)
        fun navigateById(fragmentId: Int, bundle: Bundle)
        fun navigateByDirections(navDirections: NavDirections)
    }

    companion object {
        private val DIFF_CALLBACK: DiffUtil.ItemCallback<ScreenItem> =
            object : DiffUtil.ItemCallback<ScreenItem>() {
                override fun areItemsTheSame(
                    oldItem: ScreenItem, newItem: ScreenItem
                ): Boolean {
                    return oldItem.name == newItem.name
                }

                override fun areContentsTheSame(
                    oldItem: ScreenItem, newItem: ScreenItem
                ): Boolean {
                    return oldItem == newItem
                }
            }
    }
}
